fuse/dl/cross_validation/pipeline.py

```diff
@@ -0,0 +1,303 @@
+from typing import Dict, Callable
+from fuse.utils import gpu as FuseUtilsGPU
+from sklearn.model_selection import KFold
+from functools import partial
+from multiprocessing import Process, Queue
+from typing import Sequence, Union
+import os
+from fuse.eval.metrics.classification.metrics_ensembling_common import MetricEnsemble
+from collections import OrderedDict
+from fuse.eval.evaluator import EvaluatorDefault
+from fuse.utils.file_io.file_io import create_or_reset_dir
+from fuse.utils.rand.seed import Seed
+
+
+def ensemble(test_dirs: Sequence[str], test_infer_filename: str, target_key: str, ensembled_output_file: str) -> None:
+    ensembled_output_dir = os.path.dirname(ensembled_output_file)
+    create_or_reset_dir(ensembled_output_dir, force_reset=True)
+    test_infer_filenames = [os.path.join(d, test_infer_filename) for d in test_dirs]
+    # define data for ensemble metric
+    data = {str(k): test_infer_filenames[k] for k in range(len(test_infer_filenames))}
+
+    # list of metrics
+    metrics = OrderedDict(
+        [
+            (
+                "ensemble",
+                MetricEnsemble(
+                    pred_keys=[str(i) + ".model.output.classification" for i in range(len(test_dirs))],
+                    target="0." + target_key,
+                    output_file=ensembled_output_file,
+                    output_pred_key="model.output.classification",
+                    output_target_key=target_key,
+                ),
+            ),
+        ]
+    )
+
+    evaluator = EvaluatorDefault()
+    _ = evaluator.eval(ids=None, data=data, metrics=metrics)
+
+
+def runner_wrapper(
+    q_resources: Queue,
+    rep_index: int,
+    deterministic_mode: bool,
+    fs: Union[Sequence[Callable], Callable],
+    *f_args: tuple,
+    **f_kwargs: dict,
+) -> None:
+    resource = q_resources.get()
+    print(f"Using GPUs: {resource}")
+    FuseUtilsGPU.choose_and_enable_multiple_gpus(len(resource), force_gpus=list(resource))
+    _ = Seed.set_seed(rep_index, deterministic_mode=deterministic_mode)
+    if isinstance(fs, Sequence):
+        for f, prev_arg, last_arg in zip(fs, f_args[-2], f_args[-1]):
+            f(*(f_args[:-2] + (prev_arg,) + (last_arg,)), **f_kwargs)
+    else:
+        f(*f_args, **f_kwargs)
+    print(f"Done with GPUs: {resource} - adding them back to the queue")
+    q_resources.put(resource)
+
+
+def train_wrapper(
+    sample_ids_per_fold: Sequence,
+    cv_index: int,
+    rep_index: int,
+    dataset_func: Callable,
+    dataset_params: dict,
+    paths: dict,
+    func: Callable,
+    params: dict,
+) -> None:
+
+    paths_train = paths.copy()
+
+    # set parameters specific to this fold:
+    paths_train["model_dir"] = os.path.join(paths["model_dir"], "rep_" + str(rep_index), str(cv_index))
```

Comment:
We need to specify it in our assumption that paths include "model_dir" and "cache_dir".

Reply:
done